sympy/solvers/ode/ode.py

```diff
@@ -706,6 +706,220 @@ def _helper_simplify(eq, hint, match, simplify=True, ics=None, **kwargs):
             rv = rv1
     return rv
 
+def dsubs(eq, trans, newvars=None, reverse=False):
+    r"""
+    Function that substitutes for variables in an Ordinary Differential Equation.
+
+    The inputs are -
+
+    ``trans``: A dictionary specifying the variable transformations
+    with the old variable as key and the transformation as value.
+
+    ``newvars``: A list of new variables in the transformation rules.
+    It is required if the number of new variables is not the same as
+    the number of old variables.
+
+    Returns
+    =======
+
+    The transformed differential equation
+
+    Examples
+    ========
+
+    A basic transformation can be achieved by -
+
+    >>> from sympy import symbols, dsubs, Function, cos
+    >>> f, g, h = symbols('f g h', cls=Function)
+    >>> x, a, t = symbols('x a t')
+    >>> eq = f(x).diff(x) + f(x)
+    >>> dsubs(eq, {x: t**2})
+    f(t**2) + Derivative(f(t**2), t)/(2*t)
+
+    Functions in the equation can be transformed too
+
+    >>> eq = f(x).diff(x, 2) + cos(x)
+    >>> dsubs(eq, {x: t**2, f(x): g(t)})
+    cos(t**2) + (Derivative(g(t), (t, 2))/(2*t)
+    - Derivative(g(t), t)/(2*t**2))/(2*t)
+    >>> dsubs(eq, {x: t, f(x): g(t)**2})
+    2*g(t)*Derivative(g(t), (t, 2)) + cos(t)
+    + 2*Derivative(g(t), t)**2
+
+    To apply a transformation with number of new variables greater
+    than number of old variables, a list of all new variables must be
+    given. For example, a transformation `x = a t` would be done as -
```

Comment:
We shouldn't refer to the additional symbols as "variables". The point is that here `x` and `t` are variables but `a` is intended to be a constant. The `dsubs` function has no way to know which of the symbols `a` and `t` is a variable and which is a constant so the user needs to specify that.

Reply:
Should I mention that its required when the number of new symbols is more than the number of old symbols?

Reply:
I would say something like: The newvars argument specifies which symbols should be considered variables rather than constants. For example a transformation `x = a t` where `a` is a constant and `t` is the new independent variable can be specified as... For convenience the `newvars` argument can be omitted although for robust usage in code it should always be provided. If the newvars argument is not provided then `dsubs` will treat all symbols in the replacement expression as variables but will raise an exception of the number of new variables is not equal to the number of variables being replaced.

The docstring examples should mostly pass the `newvars` argument because that it is how the function should be used robustly. Not giving newvars is okay as a convenience in interactive use but is mostly a bug magnet in proper code so the docs should encourage passing the argument.

Reply:
I've added the explanation, and used `newvars` wherever possible in the docstring.

Reply:
I still don't like the way this says "number of new variables greater than number of old variables". The number of variables is the same: the number of *symbols* is different and that is because some symbols represent variables and others represent constants.